### DISCUSSION TITLE
string_agg ordering in the resulted list

### DISCUSSION THREAD

**Discussion by lkordos** [duckdb/duckdb#9986] [Q&A] [answered]
Hi,

I am using the function string_agg. If I insert into a table values 1,2,3,4 (in this order) and then call string_agg I will get the result as a list [1,2,3,4] as expected.

My question is:

1/
Is the order of the resulted list guaranteed to be in the same order in which the values were inserted (i.e. resulted order based on rowid in ASC ordering)? Or can I expect that the result ordering would be different? Like [3,2,4,1] for example?

My tests always give me the correct results, but I am not sure if it is because the order in the result list is enforced, or if it just coincident.

2/ 
This syntax is working:
string_agg(srcColName order by rowid ASC, '|') res_col_name

However, if the ordering in the result is enforced, I guess it is not needed. 

Could someone explain me how the string_agg works internally? Thanks.

**Comment by Mytherin** [comment, marked as answer]
The result order is not guaranteed for bigger tables particularly when queries are run in parallel and/or `GROUP BY` is used. `ORDER BY` guarantees a specific result ordering. The correct syntax is `string_agg(srcColName, '|' order by rowid ASC)` (i.e. `ORDER BY` should come after **all** arguments to the function).

**Comment by lkordos** [comment]
Thank you for clarification!

**Comment by lkordos** [comment]
I think that the same issue can be regarding  the statement unnest(string_split(‘a|b|c|d’, '|')).

Is it guaranteed that there will be 4 rows, a,b,c,d with lowest rowid at ‘a’ and biggest rowid in the row with ‘d’?
Neither unnest nor string_split takes additional parameters. MS SQL Server can use the ordinal column for this porpose.

How can we preserve the ordinal order of tokens int the string with duckdb (not necessarily alphabetic) when splitting a string and unnesting it into new rows ?

Thanks.
